### PR TITLE
Fixing OAuth2JwtAccessTokenConverter tryCreateSignatureVerifier()

### PR DESCRIPTION
**Pull request by d-bagazhkov**

### CODE REVIEW

File: xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java [hunk 1]
```diff
@@ -68,7 +68,9 @@ private boolean tryCreateSignatureVerifier() {
         try {
             SignatureVerifier verifier = signatureVerifierClient.getSignatureVerifier();
             if (verifier != null) {
-                setVerifier(verifier);
+                synchronized (signatureVerifierClient) {
+                    setVerifier(verifier);
+                };
                 lastKeyFetchTimestamp = t;
```
Comment (pkev777) [inline xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java:74]: I'm worried, why lastKeyFetchTimestamp = t; - outside synchronized block if this is write operation too

File: xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java [hunk 2]
```diff
@@ -68,7 +68,9 @@ private boolean tryCreateSignatureVerifier() {
         try {
             SignatureVerifier verifier = signatureVerifierClient.getSignatureVerifier();
             if (verifier != null) {
-                setVerifier(verifier);
+                synchronized (signatureVerifierClient) {
```
Comment (sergeysenja1992) [inline xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java:71]: JEP 425 recommend to not use synchronized, better to use Lock

Comment (pkev777) [reply xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java:71]: Could you please explain? - just for clarification

Comment (pkev777) [reply xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java:71]: > JEP 425 recommend to not use synchronized, better to use Lock

Could you please explain? - just for clarification

**Review by sergeysenja1992 — APPROVED** [review]

File: xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java [hunk 3]
```diff
@@ -66,18 +66,26 @@ private boolean tryCreateSignatureVerifier() {
...
+            if (createSignatureVerifier(t)) {
                 return true;
             }
         } catch (Throwable ex) {
             log.error("could not get public key from OAuth2 server to create SignatureVerifier", ex);
         }
         return false;
     }
+
+    private synchronized boolean createSignatureVerifier(long t) throws Exception {
```
Comment (pkev777) [inline xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java:78]: Let's skip propagating time, just get time from currentTimeInMillis... inside method

File: xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java [hunk 4]
```diff
@@ -66,18 +66,26 @@ private boolean tryCreateSignatureVerifier() {
...
+        SignatureVerifier verifier = signatureVerifierClient.getSignatureVerifier();
+        if (verifier != null) {
+            setVerifier(verifier);
+            lastKeyFetchTimestamp = t;
```
Comment (pkev777) [inline xm-commons-security/src/main/java/com/icthh/xm/commons/security/oauth2/OAuth2JwtAccessTokenConverter.java:82]: let's mark lastKeyFetchTimestamp as volatile

**Review by pkev777 — APPROVED** [review]

**Review by RomanTeren — APPROVED** [review]

**Pull request merged**